File: palindrome.nim
import os
import tables
import strutils

# Graciously stolen from Nim's Wikipedia page
proc reverse(s: string): string =
    result = ""
    for i in countdown(high(s), 0):
        result.add s[i]

proc buildPalindrome(s: string, odds: CountTable[char], evens: CountTable[char]) =
    var
        buildStr = ""
        single   = ""
    for k,v in evens:
        buildStr.add repeat(k, v div 2)
    for k,v in odds:
        if v > 1:
            buildStr.add repeat(k, v div 2)
        else:
            single.add k

    buildStr = buildStr & single & reverse(buildStr)
    echo "A palindromic form of ", s, " is ", buildStr

proc checkForIfPossible(s: string) =
    var counts = initCountTable[char]()
    for c in s:
        counts.inc(c)

    var
        oddCounts  = initCountTable[char]()
        evenCounts = initCountTable[char]()
    for k,v in counts.pairs:
        if v mod 2 != 0:
            oddCounts.inc(k, v)
        else:
            evenCounts.inc(k, v)

    if oddCounts.len > 1:
        echo s, " cannot be made into a palindrome"
    else:
        buildPalindrome(s, oddCounts, evenCounts)

let input = paramStr(1)
if input == reverse(input):
    echo input, " is a palindrome"
else:
    checkForIfPossible(input)

File: palindrome.jl
function countFreqs(s)
    freqDict = Dict()
    for c in s
        freqDict[c] = get(freqDict, c, 0) + 1
    end
    return freqDict
end

function buildPalindrome(evens, odds)
    s = ""
    for e in evens
        for i in [1:div(e[2], 2)]
            s = string(s, e[1])
        end
    end
    single = ""
    for e in odds
        if e[2] > 1
            for i in [1:div(e[2], 2)]
                s = string(s, e[1])
            end
        else
            single = e[1]
        end
    end
    return string (s, single, reverse(s))
end

if ARGS[1] == reverse(ARGS[1])
    println("$(ARGS[1]) is a palindrome")
else
    freqDict = countFreqs(ARGS[1])
    oddsDict = filter((e1, e2) -> e2 % 2 != 0, freqDict)
    if length(oddsDict) > 1
        println("$(ARGS[1]) cannot be made into a palindrome")
    else
        evensDict = filter((e1, e2) -> e2 % 2 == 0, freqDict)
        println("A palindromic form of $(ARGS[1]) is $(buildPalindrome(evensDict, oddsDict))")
    end
end
